yi-core/src/Yi/CompletionTree.hs

```diff
@@ -1,59 +1,153 @@
+maximumBy' :: Eq a => (a -> a -> Ordering) -> [a] -> Maybe a
+maximumBy' cmp l | atleast 2 (== max') l = Nothing
+                 | otherwise = Just max'
+  where  max' = maximumBy cmp l
+         -- This short-circuits if the condition is met n times before the end of the list.
+         atleast :: Int -> (a -> Bool) -> [a] -> Bool
+         atleast 0 _ _ = True
+         atleast _ _ [] = False
+         atleast n cmp' (x:xs) | cmp' x = atleast (n - 1) cmp' xs
+                               | otherwise = atleast n cmp' xs
 
-mergeTrees :: Ord a => [CompletionTree a] -> CompletionTree a
-mergeTrees a = mergeTrees' (map sort' a)
-  where sort' = CT . sortBy (compareBy fst) . (\(CT l)->l)
+-- | Complete as much as possible without guessing.
+--
+-- Examples:
+--
+-- >>> complete $ fromList ["put","putStrLn","putStr"]
+-- ("put", fromList ["","Str","StrLn"])
+--
+-- >>> complete $ fromList ["put","putStr","putStrLn","xxx"]
+-- ("", fromList ["put","putStr","putStrLn","xxx"])
+complete :: ListLike a i => CompletionTree a -> (a, CompletionTree a)
+complete (CompletionTree ct)
+  | M.size ct == 1 = M.elemAt 0 ct
+  | otherwise = (mempty,CompletionTree ct)
 
-mergeTrees':: Ord a => [CompletionTree a] -> CompletionTree a
-mergeTrees' trees = CT $
-                    map (\x->(fst $ head x,mergeTrees $ map snd x)) $
-                    groupBy (((EQ==).).compareBy fst)  $
-                    sortBy (compareBy fst) $
-                    concatMap (\(CT x)->x) trees
+-- | Update the CompletionTree with new information.
+--
+-- Examples:
+--
+-- >>> update (fromList ["put","putStr"]) "p"
+-- fromList ["ut","utStr"]
+--
+-- >>> update (fromList ["put","putStr"]) "put"
+-- fromList ["","Str"]
+--
+-- >>> update (fromList ["put","putStr"]) "putS"
+-- fromList ["tr"]
+update :: (Ord a, ListLike a i, Eq i) => CompletionTree a -> a -> CompletionTree a
+update (CompletionTree ct) p
+  -- p is empty:
+  | mempty == p = CompletionTree ct
+  -- p is a key in the map ct:
+  | isJust one = fromJust one
+  -- a substring of p is a key in ct:
+  | isJust remaining = uncurry update $ fromJust remaining
+  -- p is a substring of a key in ct:
+  | otherwise = CompletionTree $ M.mapKeys fromJust
+                               $ M.filterWithKey (const . isJust)
+                               $ M.mapKeys (LL.stripPrefix p) ct
+  where
+    one = M.lookup p ct
+    remaining = listToMaybe . catMaybes $
+      map (\p' -> (,fromJust $ LL.stripPrefix p' p) <$> M.lookup p' ct) (tail $ LL.inits p)
 
-complete :: Eq a => CompletionTree a -> [a] -> ([a],CompletionTree a)
-complete tree    [] = ([],tree)
-complete (CT []) _  = ([],CT [])
-complete (CT level) (a:ta) = first ((:) a) $
-    case match of
-       Just m -> complete (snd m) ta
-       Nothing -> ([],CT [])
-     where match = find ((a==).fst) level
+-- | Like 'update' but with only one item.
+updateOne :: (Ord a, ListLike a i, Eq i) => CompletionTree a -> i -> CompletionTree a
```

Comment:
My thought process was: If you have autocomplete you probably receive one character at a time. So you might want a function that can update the completion tree one character at a time, but it ended up being pretty much `update . LL.singleton`.